# Data Understanding

**Objective:**
Load the Online Retail dataset and perform an initial inspection to understand its structure, missing values, duplicates, and general properties before proceeding to deeper Exploratory Data Analysis (EDA).

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Configuration
pd.set_option('display.max_columns', None)

## 1. Data Loading

In [2]:
file_path = '../data/raw/Online_Retail.xlsx'
df = pd.read_excel(file_path)
print("Dataset loaded successfully.")

Dataset loaded successfully.


## 2. Structure Inspection

In [3]:
print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")
df.info()

Rows: 541909, Columns: 8


<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[us]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 40.0+ MB


## 3. First Rows

In [4]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


## 4. Missing Values & Duplicates

In [5]:
missing_values = df.isnull().sum()
missing_percentages = (missing_values / len(df)) * 100
missing_df = pd.DataFrame({'Missing Count': missing_values, 'Percentage (%)': missing_percentages})
print("--- Missing Values ---")
print(missing_df[missing_df['Missing Count'] > 0])

duplicates = df.duplicated().sum()
print(f"\n--- Duplicates ---\nNumber of duplicate rows: {duplicates}")

--- Missing Values ---


             Missing Count  Percentage (%)
Description           1454        0.268311
CustomerID          135080       24.926694



--- Duplicates ---
Number of duplicate rows: 5268


## 5. Statistical Summary

In [6]:
df.describe(include='all')

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
count,541909.0,541909,540455,541909.000000,541909,541909.000000,406829.000000,541909
unique,25900.0,4070,4223,NaN,NaN,NaN,NaN,38
top,573585.0,85123A,WHITE HANGING HEART T-LIGHT HOLDER,NaN,NaN,NaN,NaN,United Kingdom
freq,1114.0,2313,2369,NaN,NaN,NaN,NaN,495478
mean,NaN,NaN,NaN,9.552250,2011-07-04 13:34:57.156386,4.611114,15287.690570,NaN
min,NaN,NaN,NaN,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000,NaN
25%,NaN,NaN,NaN,1.000000,2011-03-28 11:34:00,1.250000,13953.000000,NaN
50%,NaN,NaN,NaN,3.000000,2011-07-19 17:17:00,2.080000,15152.000000,NaN
75%,NaN,NaN,NaN,10.000000,2011-10-19 11:27:00,4.130000,16791.000000,NaN
max,NaN,NaN,NaN,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000,NaN


### Key Findings
1. **Structure**: The dataset contains 541,909 rows and 8 columns.
2. **Data Types**: The dataset contains categorical (`InvoiceNo`, `StockCode`, `Description`, `Country`), numerical (`Quantity`, `UnitPrice`, `CustomerID`), and temporal (`InvoiceDate`) features.
3. **Missing Values**:
    * `CustomerID`: 135,080 missing values (~24.93%). This is a critical issue as customer segmentation requires customer identification.
    * `Description`: 1,454 missing values (~0.27%).
4. **Duplicates**: There are 5,268 duplicate rows in the dataset.
5. **Unusual Values**: There are negative quantities and potentially zero or negative UnitPrices, which indicate returns or bad records.

### Decisions Made
* Proceed to EDA to investigate the exact distributions of Quantity, UnitPrice, and the temporal patterns.
* Missing `CustomerID` records will likely need to be excluded or handled separately in the cleaning phase since they cannot be mapped to a specific customer for segmentation.
* Duplicate rows will need to be dropped in the preprocessing pipeline.